pkg/database/billing.go

```diff
@@ -0,0 +1,126 @@
+package database
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+)
+
+type billing struct {
+	c    cosmosdb.BillingDocumentClient
+	uuid string
+}
+
+// Billing is the database interface for BillingDocuments
+type Billing interface {
+	Create(context.Context, *api.BillingDocument) (*api.BillingDocument, error)
+	Get(context.Context, string) (*api.BillingDocument, error)
+	Patch(context.Context, string, func(*api.BillingDocument) (bool, error)) (*api.BillingDocument, error)
+}
+
+// NewBilling returns a new Billing
+func NewBilling(ctx context.Context, uuid string, dbc cosmosdb.DatabaseClient, dbid, collid string) (Billing, error) {
+	collc := cosmosdb.NewCollectionClient(dbc, dbid)
+
+	triggers := []*cosmosdb.Trigger{
+		{
+			ID:               "setCreationBillingTimeStamp",
+			TriggerOperation: cosmosdb.TriggerOperationAll,
+			TriggerType:      cosmosdb.TriggerTypePre,
+			Body: `function trigger() {
+	var request = getContext().getRequest();
+	var body = request.getBody();
+	var date = new Date();
+	var now = Math.floor(date.getTime() / 1000);
+	var billingBody = body["billing"];
+	billingBody["creationTime"] = now;
+	billingBody["lastBillingTime"] = now;
+	request.setBody(body);
+}`,
+		},
+		{
+			ID:               "setDeletionBillingTimeStamp",
+			TriggerOperation: cosmosdb.TriggerOperationAll,
+			TriggerType:      cosmosdb.TriggerTypePre,
+			Body: `function trigger() {
+	var request = getContext().getRequest();
+	var body = request.getBody();
+	var date = new Date();
+	var now = Math.floor(date.getTime() / 1000);
+	var billingBody = body["billing"];
+	billingBody["deletionTime"] = now;
+	request.setBody(body);
+}`,
+		},
+	}
+
+	triggerc := cosmosdb.NewTriggerClient(collc, collid)
+	for _, trigger := range triggers {
+		_, err := triggerc.Create(ctx, trigger)
+		if err != nil && !cosmosdb.IsErrorStatusCode(err, http.StatusConflict) {
+			return nil, err
+		}
+	}
+
+	return &billing{
+		c:    cosmosdb.NewBillingDocumentClient(collc, collid),
+		uuid: uuid,
+	}, nil
+}
+
+// Creating Billing Document
+func (c *billing) Create(ctx context.Context, doc *api.BillingDocument) (*api.BillingDocument, error) {
```

Comment:
please can you split this out so that the create/get/patch functions are vanilla and very similar to the functionality for other collections (the functions can be unexported to prevent them being called accidentally) and add an intent-based API (e.g. "MarkDeleted()" or whatever) which is opinionated on how it calls create(), patch(), etc?

this will make future unit testing easier as well as removing the need for "// Returns true means we will update deletion timestamp"

Reply:
I removed the `bool` inside the Patch operation.

I guess this makes the Create/Get/Patch back to normal operations if i compare to the other one.

Is it ok like this or am i missing something ?

Reply:
It'd be better if calling Patch didn't trigger setDeletionBillingTimeStamp.  It could cause an ugly mistake later.  Either rename Patch, or (preferred) split the function into two pieces, one `patch(ctx context.Context, key string, f func(*api.BillingDocument) error, options *cosmosdb.Options) (*api.BillingDocument, error)` and another `MarkDeleted(ctx context.Context, key string)` which calls patch()